pandas/core/algorithms.py

```diff
@@ -944,14 +944,27 @@ def select_n_frame(frame, columns, n, method, keep):
     -------
     nordered : DataFrame
     """
-    from pandas.core.series import Series
     if not is_list_like(columns):
         columns = [columns]
-    columns = list(columns)
-    ser = getattr(frame[columns[0]], method)(n, keep=keep)
-    if isinstance(ser, Series):
-        ser = ser.to_frame()
-    return ser.merge(frame, on=columns[0], left_index=True)[frame.columns]
+    else:
+        columns = list(columns)
+    reverse = method == 'nlargest'
+    for i, column in enumerate(columns):
+        series = frame[column]
+        if reverse:
```

Comment:
doesn't this defeat the purpose of ``.nlargest()``? e.g. you are sorting the entire series.

why not just use ``series.nlargest(n)`` here? (or ``.nsmallest``). And defer to the duplicate handling there (I *think* that works).